Source/WebCore/platform/text/WritingMode.h

```diff
@@ -31,233 +31,397 @@
+    constexpr inline bool isLineOverRight() const; // line-over == right
+    constexpr inline bool isLineOverLeft() const;  // line-over == left
+
+    // Coordinate flow queries.
+    constexpr inline bool isLogicalLeftLineLeft() const; // Not true for sideways-lr.
+    constexpr inline bool isLogicalLeftInlineStart() const; // == !isInlineFlipped()
+    constexpr inline bool isInlineFlipped() const; // Inline direction is RTL or BTT.
+    constexpr inline bool isBlockFlipped() const;  // Block direction is RL or BT.
+
+    // Directions as enums. Prefer booleans if doing boolean checks.
+    constexpr inline FlowDirection blockDirection() const;
+    constexpr inline TextDirection bidiDirection() const;
+    constexpr inline FlowDirection inlineDirection() const;
+
+    // Computed values. May differ from used values above.
+    constexpr inline WritingModeStyle computedWritingMode() const;
+    constexpr inline TextDirection computedTextDirection() const;
+    constexpr inline TextOrientation computedTextOrientation() const;
+    // ^ USE THESE FOR STYLE COMPUTATION ^ //
+
+    // Setters.
+    inline void setWritingMode(WritingModeStyle);
+    inline void setTextOrientation(TextOrientation);
+    inline void setTextDirection(TextDirection);
+
+    // To aid in packing.
+    using Data = uint8_t;
+    Data toData() const { return m_bits; }
+    constexpr inline WritingModeData(Data bits = 0) { m_bits = bits; }
+
+    friend bool operator==(WritingModeData, WritingModeData) = default;
+
+private:
+    Data m_bits { 0 };
+
+public: // Private except WritingModeStyle and FlowDirection are friends
+    enum Bits {
+        kIsVerticalText  = 1 << 0, // Vertical writing modes.
+        kIsFlippedBlock  = 1 << 1, // RL or BT block flow directions.
+        kIsVerticalType  = 1 << 2, // Vertical typographic mode.
+        kIsBidiRTL       = 1 << 3, // Bidi directionality.
+        kIsUprightType   = 1 << 4, // Upright text orientation.
+        kIsSidewaysType  = 1 << 5, // Sideways text orientation.
+
+        kBlockFlowMask   = kIsVerticalText | kIsFlippedBlock,
+        kWritingModeMask = kBlockFlowMask  | kIsVerticalType,
+        kOrientationMask = kIsSidewaysType | kIsUprightType, // Both is an error.
+        kOrientationShift = 4,
+    };
 };
 
-enum class WritingMode : uint8_t {
-    HorizontalTb,
-    HorizontalBt, // Non-standard
-    VerticalLr,
-    VerticalRl,
-    SidewaysLr,
-    SidewaysRl,
+enum class WritingModeStyle : uint8_t {
+    HorizontalTb = 0,
+    HorizontalBt = WritingModeData::kIsFlippedBlock, // Non-standard.
+    VerticalLr   = WritingModeData::kIsVerticalText | WritingModeData::kIsVerticalType,
+    VerticalRl   = WritingModeData::kIsVerticalText | WritingModeData::kIsVerticalType | WritingModeData::kIsFlippedBlock,
+    SidewaysLr   = WritingModeData::kIsVerticalText,
+    SidewaysRl   = WritingModeData::kIsVerticalText | WritingModeData::kIsFlippedBlock,
 };
```

Comment:
I suppose the goal here is to allow using static_casts and minimize conversions when constructing `WritingModeData`?

Reply:
And when spitting `WritingModeStyle` and `FlowDirection` values back out of the `WritingModeData` object.